In [81]:
#Import Libraries
import pandas as pd
import numpy as np 
import os

In [7]:
#Create path
path= '/Users/shantamaraj/Desktop/INSTACART DATA'
path

'/Users/shantamaraj/Desktop/INSTACART DATA'

In [14]:
#Import original products.csv and wrangled orders.csv
df_prods = pd.read_csv(os.path.join(path, 'products.csv'), index_col= False)
df_ords = pd.read_csv(os.path.join(path, 'orders_wrangled.csv'), index_col= False)

In [16]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [17]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [18]:
#Reset index
df_ords = df_ords.reset_index(drop=True)

In [19]:
#Drop column
if 'Unnamed: 0' in df_ords.columns:
    df_ords = df_ords.drop(columns=['Unnamed: 0'])

In [20]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [21]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [22]:
#Create a dataframe
df_test= pd.DataFrame()

In [23]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [24]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [25]:
#Check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis =1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [26]:
#Change mixed data type to 'string'
df_test['mix'] = df_test['mix'].astype('str')

In [27]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [28]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [29]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [30]:
df_prods.shape

(49693, 5)

In [31]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [32]:
df_prods_clean.shape

(49677, 5)

In [33]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [34]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [35]:
df_prods_clean.shape

(49677, 5)

In [36]:
#Drop duplicate data
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [37]:
df_prods_clean_no_dups.shape

(49672, 5)

In [38]:
#Export checked data frame
df_prods_clean_no_dups.to_csv(os.path.join(path, 'products_checked.csv'))

TASK 4.5

In [40]:
#2 run df_ords.describe()
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#The data contains a large number of observations (approx 3.4 million)
'order_id', 'user_id' and 'order_number' are identifier variables and the data type should be considered 'string'
'orders_day_of_week' has a minimum of '0' and a maximum of '6' suggesting the days are zero-indexed (0 for Monday or Sunday)
The minimum value for 'days_since_prior_order' is 0, suggesting multiple orders on the same day or a placeholder for customers' first orders
The count for 'days_since_prior_order' is also lower compared to the other columns, suggesting missing data

In [43]:
#3 check for mixed-type data in your df_ords dataframe
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis =1)
    if len (df_test[weird]) > 0:
        print(col)

In [46]:
#4 There is no mixed-type data in the df_ords data frame

In [44]:
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [64]:
#5 Run a check for missing values in df_ords data frame
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#There are 206,209 missing values in the 'days_since_prior_order' column.

In [60]:
#6 Address missing values using an apropriate method
# Create subset of the dataframe that contains the nulls
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [61]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [62]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [65]:
#There is no need to remove or alter the missing values. All NAN values from "days_since_prior_order" corresponds with 
#customer's first order. These are natrually missing because there is no prior order to compare agaisnt.
#Removing or imputing the data could lead to distorted or misleading data.
#The missingness in this context is meaningful (represents first orders)

In [67]:
#7 Run a check for duplicate values in df_ords data
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [68]:
#There are no duplicates in the df_ords dataframe. 

In [69]:
#8 Address the duplicates using an appropriate method - N/A

In [70]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [71]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [79]:
#Export clean data as csv file
df_ords.to_csv(os.path.join(path, 'orders_clean.csv'))

In [82]:
df_prods.to_csv(os.path.join(path, 'products_clean.csv'))